In [1]:
import numpy as np
from pandas import DataFrame, read_csv
# read_csv()
import os

BLOCK_SIZE = 20
FILTER_THRESHOLD = 4

In [2]:
from collections import namedtuple, defaultdict


# Point = namedtuple('Point',['window', 'snp', 'bp'])

class Point:
    def __init__(self, snp, bp=0):
        self.snp, self.bp = snp, bp
    
    @property
    def window(self):
        return self.snp // BLOCK_SIZE

# class Specimen:
#     def __init__(self, ident, sequence)
#         ident, sequence
    
class Node:
    def __init__(self, ident, start, end, upstream=None, downstream=None):
        self.ident = ident
        self.start = start #Point()
        self.end = end #Point()
        # {nothing_node:501, Node: 38,  Node: 201, Node: 3}
        self.upstream = defaultdict(lambda: 0) if not upstream else upstream 
        # {Node: 38,  Node: 201, Node: 3}
        self.downstream = defaultdict(lambda: 0) if not downstream else downstream
        self.specimens = set()
    
    def __len__(self):
        return len(self.specimens)
    
    def __repr__(self):
        return "N%s(%s, %s)" % (str(self.ident), str(self.start.snp), str(self.end.snp))
    
    def __hash__(self):
        return hash(self.ident) + hash(self.start.snp) + hash(self.end.snp)
    
    def details(self):
        return f"""Node{self.ident}: {self.start.snp} - {self.end.snp}
        upstream: { dict((key, value) for key,value in self.upstream.items()) }
        downstream: { dict((key, value) for key,value in self.downstream.items()) }
        specimens: {self.specimens}"""
        

a = Point(0)
b = Point(14)
str(Node(57, a, b))
nothing_node = Node(-1, Point(None), Point(None))
global_nodes = {0: nothing_node}


deepcopy(a)

In [3]:
def read_data(file_path = "../test_data/KE_chromo10.txt"):
    """Individuals are rows, not columns"""
    loci = []
    with open(file_path) as ke:
        for line in ke.readlines():
            loci.append(tuple(int(x) for x in line.split()))
            
    
    individuals = np.array(loci).T.tolist()
    return loci, individuals
alleles, individuals = read_data()
assert len(alleles) == 32767
assert len(individuals[1]) == 32767
assert len(individuals) == 501

In [4]:
def first(iterable):
    return next(iter(iterable))

In [5]:
def signature(individual, start_locus):
    return tuple(individual[start_locus : start_locus + BLOCK_SIZE])

def get_unique_signatures(individuals, start_locus, block_size = 20):
    unique_blocks = {}
    for individual in individuals:
        sig = signature(individual, start_locus)
        if sig not in unique_blocks:
            unique_blocks[sig] = Node(len(unique_blocks), Point(start_locus // block_size, start_locus), 
                                      Point(start_locus // block_size, start_locus + BLOCK_SIZE)) #TODO: -1?
    
    return unique_blocks
unique_blocks = get_unique_signatures(individuals, 0 )
    
assert len(unique_blocks) == 4
unique_blocks
# assert unique_blocks == {(0, 2, 0, 0, 2, 0, 2, 0, 0, 0, 2, 0, 0, 0, 2, 2, 0, 0, 0, 0): 0,
#  (0, 0, 2, 2, 0, 2, 0, 2, 2, 2, 0, 2, 2, 2, 0, 0, 2, 2, 2, 2): 1,
#  (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0): 2,
#  (2, 0, 2, 2, 0, 2, 0, 2, 2, 2, 0, 2, 2, 2, 0, 0, 2, 2, 2, 2): 3}

{(0, 2, 0, 0, 2, 0, 2, 0, 0, 0, 2, 0, 0, 0, 2, 2, 0, 0, 0, 0): N0(0, 0),
 (0, 0, 2, 2, 0, 2, 0, 2, 2, 2, 0, 2, 2, 2, 0, 0, 2, 2, 2, 2): N1(0, 0),
 (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0): N2(0, 0),
 (2, 0, 2, 2, 0, 2, 0, 2, 2, 2, 0, 2, 2, 2, 0, 0, 2, 2, 2, 2): N3(0, 0)}

In [6]:
def get_all_signatures(alleles, individuals):
    unique_signatures = []
    for locus_start in range(0, len(alleles) - BLOCK_SIZE, BLOCK_SIZE):  # discards remainder 
        sig = get_unique_signatures(individuals, locus_start, BLOCK_SIZE)
        unique_signatures.append(sig)
    return unique_signatures
unique_signatures = get_all_signatures(alleles, individuals)

In [7]:
unique_signatures[21]

{(0, 0, 0, 0, 2, 2, 0, 2, 0, 2, 2, 2, 2, 2, 2, 2, 0, 2, 0, 2): N0(21, 21),
 (0, 0, 2, 0, 2, 2, 0, 2, 0, 2, 2, 2, 2, 2, 2, 2, 0, 2, 0, 2): N1(21, 21),
 (0, 0, 0, 2, 2, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0): N2(21, 21),
 (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0): N3(21, 21),
 (0, 0, 0, 0, 2, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0): N4(21, 21),
 (0, 0, 0, 2, 2, 2, 0, 2, 0, 2, 2, 2, 2, 0, 0, 2, 0, 0, 0, 2): N5(21, 21),
 (0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0): N6(21, 21),
 (0, 0, 0, 0, 2, 2, 0, 2, 2, 2, 2, 2, 2, 0, 0, 2, 0, 0, 2, 2): N7(21, 21)}

In [8]:
def build_individuals(individuals, unique_signatures):
    simplified_individuals = []
    for i_specimen, specimen in enumerate(individuals):
        my_simplification = []
        for w, window in enumerate(unique_signatures):  # the length of the genome
            sig = signature(specimen, w * BLOCK_SIZE)
    #         print(sig, unique_signatures[w][sig])
    #         print(i_specimen, window)
            my_simplification.append(unique_signatures[w][sig])
        simplified_individuals.append(my_simplification)
    return simplified_individuals
simplified_individuals = build_individuals(individuals, unique_signatures)
print(simplified_individuals[500][:100])
len(simplified_individuals), len(simplified_individuals[60])

[N2(0, 0), N2(1, 1), N2(2, 2), N2(3, 3), N2(4, 4), N2(5, 5), N3(6, 6), N3(7, 7), N3(8, 8), N2(9, 9), N0(10, 10), N1(11, 11), N2(12, 12), N2(13, 13), N2(14, 14), N2(15, 15), N3(16, 16), N3(17, 17), N4(18, 18), N3(19, 19), N5(20, 20), N3(21, 21), N3(22, 22), N10(23, 23), N4(24, 24), N3(25, 25), N4(26, 26), N3(27, 27), N1(28, 28), N1(29, 29), N4(30, 30), N3(31, 31), N21(32, 32), N1(33, 33), N1(34, 34), N1(35, 35), N1(36, 36), N1(37, 37), N1(38, 38), N1(39, 39), N1(40, 40), N1(41, 41), N1(42, 42), N1(43, 43), N1(44, 44), N1(45, 45), N1(46, 46), N1(47, 47), N1(48, 48), N1(49, 49), N1(50, 50), N1(51, 51), N1(52, 52), N1(53, 53), N1(54, 54), N1(55, 55), N1(56, 56), N1(57, 57), N1(58, 58), N1(59, 59), N1(60, 60), N1(61, 61), N1(62, 62), N1(63, 63), N1(64, 64), N1(65, 65), N1(66, 66), N1(67, 67), N1(68, 68), N1(69, 69), N1(70, 70), N1(71, 71), N1(72, 72), N1(73, 73), N1(74, 74), N1(75, 75), N1(76, 76), N1(77, 77), N0(78, 78), N0(79, 79), N1(80, 80), N1(81, 81), N1(82, 82), N1(83, 83), N1(84, 84

(501, 1638)

# Nodes: Populate upstream and downstream

In [9]:
# build nodes:  first 4 are the 4 starting signatures in window 0.  
# For each node list which individuals are present at that node
# List transition rates from one node to all other upstream and downstream
def populate_transitions(simplified_individuals):
    for i, indiv in enumerate(simplified_individuals):
        # look what variants are present
        for x, node in enumerate(indiv):
            node.specimens.add(i)
            if x + 1 < len(indiv):
                node.downstream[indiv[x+1]] += 1
            else:
                node.downstream[nothing_node] += 1
            if x-1 >= 0:
                node.upstream[indiv[x-1]] += 1
            else: 
                node.upstream[nothing_node] += 1
            

In [10]:
unique_signatures = get_all_signatures(alleles, individuals)
simplified_individuals = build_individuals(individuals, unique_signatures)
populate_transitions(simplified_individuals)

#### TODO: turn these into tests

In [11]:
simplified_individuals[50][0].downstream

defaultdict(<function __main__.Node.__init__.<locals>.<lambda>()>,
            {N1(1, 1): 286})

In [12]:
simplified_individuals[49][0].downstream

defaultdict(<function __main__.Node.__init__.<locals>.<lambda>()>,
            {N0(1, 1): 103})

In [13]:
simplified_individuals[500][0].downstream

defaultdict(<function __main__.Node.__init__.<locals>.<lambda>()>,
            {N2(1, 1): 82})

In [14]:
simplified_individuals[91][0].downstream

defaultdict(<function __main__.Node.__init__.<locals>.<lambda>()>,
            {N1(1, 1): 30})

In [15]:
[x.downstream.values() for x in unique_signatures[1000].values()]

[dict_values([299]), dict_values([120]), dict_values([82])]

In [16]:
[x.upstream.values() for x in unique_signatures[1000].values()]

[dict_values([102, 197]), dict_values([120]), dict_values([82])]

---------------

# Simple Merge

In [17]:
# TODO: add signature directly to node

In [18]:
from blist import blist
from copy import copy, deepcopy

In [19]:
def test_no_duplicate_nodes(global_nodes):
    unique_nodes = set()
    for node in global_nodes:
        if node in unique_nodes:
            print(node)
        else:
            unique_nodes.add(node)


In [20]:
# zoom_stack = [[]]
def simple_merge(global_nodes):
    new_layer = []  # TODO: copy old nodes to new layer conditionally
    n = 0
    while n < len(global_nodes):  # size of global_nodes changes, necessitating this weird loop
        node = global_nodes[n]
    #     print(node, type(node))
        if len(node.downstream) == 1: 
            next_node = first(node.downstream.keys())
            if len(node.specimens) == len(next_node.specimens):
                #Torsten deletes nodeA and modifies next_node
                next_node.upstream = node.upstream
                next_node.start = node.start
                #prepare to delete node by removing references
                for parent in node.upstream.keys():
                    if parent != nothing_node:
                        count = parent.downstream[node]
                        del parent.downstream[node]  # updating pointer 
                        parent.downstream[next_node] = count 
                global_nodes.remove(node)  #delete node
                # zoom_stack[0].append(merged)
                n -= 1
        n += 1
    return global_nodes        

In [21]:
def test_simple_merge():
    global_nodes = blist([node for window in unique_signatures for node in window.values()])  # think about referencing and deletion
    assert len(global_nodes) == 7180
    summary1 = simple_merge(global_nodes)
    assert len(summary1) == 3690
    return summary1
summary1 = test_simple_merge()

#### Neglect Nodes

In [22]:
def delete_node(node, cutoff):
    """Changes references to this node to add to references to nothing_node"""
    if cutoff < 1:
        return  # if cutoff is 0, then don't touch upstream and downstream
    for parent, count in node.upstream.items():
        parent.downstream[nothing_node] += parent.downstream[node]
        del parent.downstream[node]
    for descendant, count in node.downstream.items():
        descendant.upstream[nothing_node] += descendant.upstream[node]
        del descendant.upstream[node]
        

def neglect_nodes(all_nodes, deletion_cutoff=FILTER_THRESHOLD):
    nodes_to_delete = set()
#     filtered_nodes = copy(all_nodes)
#     filtered_nodes.remove(1)
#     assert len(all_nodes) != len(filtered_nodes)
    for node in all_nodes:
        if len(node.specimens) <= deletion_cutoff:
            delete_node(node, deletion_cutoff)  # TODO: check if this will orphan 
            nodes_to_delete.add(node)
    filtered_nodes = blist([x for x in all_nodes if x not in nodes_to_delete])
    # TODO: remove orphaned haplotypes in a node that transition to and from zero within a 10 window length
    return filtered_nodes 


def test_neglect_nodes(all_nodes):
    summary2 = neglect_nodes(all_nodes)
    assert len(summary2) == 2854
    unchanged = neglect_nodes(summary2, 0)
    assert len([n for n in unchanged if len(n.specimens) == 0]) == 0
    return summary2
summary2 = test_neglect_nodes(summary1)

In [23]:
summary2[8].details()

'Node2: 7 - 7\n        upstream: {N2(6, 6): 158, N-1(None, None): 2}\n        downstream: {N2(8, 8): 158, N3(8, 8): 2}\n        specimens: {4, 5, 8, 14, 17, 26, 29, 35, 37, 38, 53, 59, 61, 62, 65, 71, 77, 79, 86, 90, 92, 100, 101, 102, 105, 108, 111, 112, 113, 115, 117, 119, 121, 123, 124, 125, 126, 129, 133, 134, 137, 138, 140, 142, 151, 155, 162, 163, 167, 169, 175, 176, 177, 180, 181, 182, 183, 186, 190, 191, 193, 199, 202, 210, 212, 216, 219, 221, 224, 228, 229, 231, 233, 236, 240, 243, 245, 246, 249, 252, 254, 256, 258, 263, 264, 268, 272, 276, 278, 284, 287, 290, 292, 293, 297, 299, 300, 301, 302, 304, 305, 308, 309, 312, 319, 326, 327, 332, 333, 334, 337, 338, 339, 346, 348, 352, 354, 357, 358, 359, 360, 364, 365, 369, 378, 379, 382, 388, 394, 395, 396, 399, 406, 407, 408, 409, 414, 417, 421, 424, 426, 427, 435, 436, 440, 449, 452, 457, 465, 467, 468, 480, 484, 485, 486, 487, 490, 492, 495, 498}'

#### Split Groups

In [24]:
def split_one_group(prev_node, anchor, next_node):
    """ Called when up.specimens == down.specimens"""
    new = Node(777, prev_node.start, next_node.end, prev_node.upstream, next_node.downstream)  # TODO: what about case where more content is joining downstream?
    # Comment: That is actually the case we want to split up to obtain longer blocks later
    # Extension of full windows will take care of potential loss of information later
    
    if nothing_node != prev_node:
        new.specimens = anchor.specimens.intersection(prev_node.specimens)
    elif nothing_node != next_node:
        new.specimens = anchor.specimens.intersection(next_node.specimens)
    else:
        new.specimens = anchor.specimens
        for n in next_node.downstream.keys():
            if n != nothing_node:
                new.specimens = new.specimens.remove(n.specimens)
        for n in prev_node.upstream.keys():
            if n != nothing_node:
                new.specimens = new.specimens.remove(n.specimens)
    
    if nothing_node is prev_node:  # Rare case
        new.start = anchor.start
        new.upstream = anchor.upstream
    if nothing_node is next_node:
        new.end = anchor.end
        new.downstream = anchor.downstream
        

    
    print(new.details())
    print(new.upstream.keys())
    print(new.upstream.values())
    print(sum(new.upstream.values()))
    
    # Update upstream/downstream
    running = new.upstream.keys()

    ## n.upstream/downstream contains the same key multiple times?!
    ## My quick fix was to delete all upstream/downstream and just recalculate everything...
    new.upstream = defaultdict(lambda: 0) 
    for n in running:
        if n != nothing_node:
            new.upstream[n] = len(new.specimens.intersection(n.specimens))
            n.downstream[new] = new.upstream[n]
            n.downstream[prev_node] = n.downstream[prev_node] - n.downstream[new]
            if n.downstream[prev_node] == 0:
                del n.downstream[prev_node]
    
    running = new.downstream.keys()
    new.downstream = defaultdict(lambda: 0)
    for n in running:
        if n != nothing_node:
            new.downstream[n] = len(new.specimens.intersection(n.specimens))
            n.upstream[new] = new.downstream[n]
            n.upstream[next_node] = n.upstream[next_node] - n.upstream[new]
            if n.upstream[next_node] == 0:
                del n.upstream[next_node]
    
    print(new.details())
    print(new.upstream.keys())
    print(new.upstream.values())
    print(sum(new.upstream.values()))
    
    accounted_upstream = sum(new.upstream.values()) - new.upstream[nothing_node]
    #print(f'upstream {sum(new.upstream.values())} downstream {sum(new.downstream.values())}')
    new.upstream[nothing_node] = len(new.specimens) - accounted_upstream
    accounted_downstream = sum(new.downstream.values()) - new.downstream[nothing_node]
    new.downstream[nothing_node] = len(new.specimens) - accounted_downstream 
    
    assert all([count > -1 for count in new.upstream.values()]), new.details()
    assert all([count > -1 for count in new.downstream.values()]), new.details()
    # Update Specimens in prev_node, anchor, next_node
    if prev_node != nothing_node:
        prev_node.specimens -= new.specimens
    
    if next_node != nothing_node:
        next_node.specimens -= new.specimens
    
    anchor.specimens -= new.specimens
        
    ## anchor.specimens.difference_update(prev_node.specimens) REASON?
    return new

test_graph = summary2  # deepcopy(
example = test_graph[7]
original = deepcopy(example)
print(example.details())
def test_split_one_group(prev_node, anchor, next_node):
    x = split_one_group(prev_node, anchor, next_node)
    assert x
    answer = set(int(x)-1 for x in '14  16  19  20  28  56  59  69  88 133 140 155 159 160 175 193 199 201 224 249 252 258 260 267 268 283 292 295 318 322 325 332 341 344 346 351 354 357 362 364 367 373 374 375 381 386 392 393 394 402 403 417 421 424 426 431 434 435 438 442 445 447 452 455 457 462 463 464 467 471 473 475 476 477 478 480 483 484 494 497 501'.split())
    assert x.specimens == answer, 'Specimens set does not agree with HaploBlocker' + str(x.specimens.difference(answer))
    return x

x = test_split_one_group(first(example.upstream),  example, first(example.downstream) )

example.details()

original[7].details()

In [25]:
def split_groups(all_nodes):
    """This is called crossmerge in the R code"""
    number_of_windows = len(first(simplified_individuals))
    length = len(all_nodes)# size of global_nodes changes, necessitating this weird loop
    for n in range(length):  
        node = all_nodes[n]
        #check if all transitition upstream match with one of my downstream nodes
        #if set(node.upstream.values()) == set(node.downstream.values()): WHY?
        if node.start.snp != 0 and node.end.window != number_of_windows: #chr begin or end
            if len(node.specimens) > 0:
                # Matchup upstream and downstream with specimen identities
                for up in tuple(node.upstream.keys()):
                    for down in tuple(node.downstream.keys()):

                        set1 = up.specimens
                        set2 = down.specimens
                        if up == nothing_node:
                            set1 = node.specimens
                            for index in tuple(node.upstream.keys()):
                                set1.intersection(index.specimens) # =- does not work for empty sets
                        if down == nothing_node:
                            set2 = node.specimens
                            for index in tuple(node.downstream.keys()):
                                set2.intersection(index.specimens) # =- does not work for empty sets

                        if set1 == set2 and len(set1) > 0:
                            new_node = split_one_group(up, node, down)
                            all_nodes.append(new_node)
        
    filtered = neglect_nodes(all_nodes, 0)
    return filtered
    

def test_split_groups(all_nodes):
    summary3 = split_groups(all_nodes)
    assert summary3
    return summary3
summary3 = test_split_groups(summary2)
    

Node777: 0 - 7
        upstream: {N-1(None, None): 82}
        downstream: {N3(8, 8): 82}
        specimens: {257, 385, 259, 132, 391, 392, 393, 266, 139, 267, 13, 15, 401, 18, 19, 402, 282, 154, 27, 158, 159, 416, 291, 420, 294, 423, 425, 174, 430, 433, 434, 437, 55, 441, 58, 444, 317, 446, 192, 321, 451, 68, 324, 198, 454, 200, 456, 331, 461, 462, 463, 466, 340, 470, 87, 343, 345, 472, 474, 475, 476, 350, 223, 477, 353, 479, 482, 356, 483, 361, 363, 493, 366, 496, 372, 373, 374, 500, 248, 251, 380}
dict_keys([N-1(None, None)])
dict_values([82])
82
Node777: 0 - 7
        upstream: {}
        downstream: {N3(8, 8): 81}
        specimens: {257, 385, 259, 132, 391, 392, 393, 266, 139, 267, 13, 15, 401, 18, 19, 402, 282, 154, 27, 158, 159, 416, 291, 420, 294, 423, 425, 174, 430, 433, 434, 437, 55, 441, 58, 444, 317, 446, 192, 321, 451, 68, 324, 198, 454, 200, 456, 331, 461, 462, 463, 466, 340, 470, 87, 343, 345, 472, 474, 475, 476, 350, 223, 477, 353, 479, 482, 356, 483, 361, 363, 493, 36

148
Node777: 420 - 448
        upstream: {N777(413, 419): 148, N777(413, 419): 0}
        downstream: {N0(449, 449): 148}
        specimens: {0, 9, 14, 20, 25, 27, 30, 34, 36, 37, 41, 44, 48, 49, 58, 60, 62, 63, 82, 94, 98, 106, 108, 116, 119, 121, 127, 137, 139, 150, 181, 189, 208, 211, 218, 219, 226, 228, 229, 233, 234, 235, 237, 238, 242, 246, 247, 252, 259, 262, 265, 267, 273, 274, 276, 280, 283, 284, 285, 287, 288, 295, 300, 307, 316, 319, 320, 321, 322, 324, 327, 329, 330, 332, 333, 334, 337, 338, 339, 345, 347, 352, 356, 357, 358, 359, 361, 364, 365, 371, 372, 374, 378, 379, 380, 384, 387, 388, 389, 397, 398, 399, 400, 401, 402, 404, 405, 406, 412, 415, 416, 421, 422, 426, 427, 428, 437, 438, 439, 440, 441, 444, 446, 447, 449, 450, 453, 455, 456, 457, 459, 465, 468, 472, 473, 474, 476, 477, 479, 480, 484, 486, 487, 491, 492, 494, 496, 497}
dict_keys([N777(413, 419), N777(413, 419)])
dict_values([148, 0])
148
Node777: 420 - 448
        upstream: {N777(413, 419): 0, N777(413, 419)

dict_values([0])
0
Node777: 769 - 769
        upstream: {N1(765, 768): 1, N-1(None, None): 2, N777(757, 768): 30, N777(765, 768): 186}
        downstream: {N2(770, 770): 187, N3(770, 771): 30, N-1(None, None): 2}
        specimens: {4, 8, 9, 11, 12, 13, 15, 16, 17, 18, 19, 21, 27, 29, 30, 35, 39, 40, 42, 51, 53, 55, 56, 58, 68, 71, 75, 77, 79, 80, 83, 86, 90, 95, 97, 105, 107, 111, 112, 114, 115, 124, 125, 128, 131, 133, 134, 138, 141, 142, 146, 151, 154, 155, 158, 159, 164, 168, 170, 171, 172, 173, 175, 176, 178, 180, 181, 183, 186, 187, 190, 191, 192, 194, 198, 200, 202, 204, 205, 216, 220, 221, 223, 227, 232, 236, 243, 245, 251, 253, 254, 256, 259, 263, 264, 266, 269, 270, 271, 273, 278, 279, 286, 289, 291, 293, 297, 301, 304, 305, 306, 308, 310, 312, 313, 314, 315, 316, 319, 324, 326, 328, 330, 331, 332, 334, 335, 338, 339, 341, 342, 344, 345, 346, 347, 349, 352, 355, 358, 360, 362, 365, 367, 369, 370, 374, 375, 376, 377, 381, 382, 383, 387, 388, 389, 390, 391, 393, 394, 396, 398, 

164
Node777: 1189 - 1206
        upstream: {N4(1186, 1188): 25, N-1(None, None): 0}
        downstream: {N2(1207, 1207): 25}
        specimens: {137, 267, 397, 405, 279, 409, 154, 155, 159, 287, 417, 422, 423, 300, 307, 191, 320, 332, 346, 219, 226, 228, 236, 370, 371}
dict_keys([N4(1186, 1188), N-1(None, None)])
dict_values([25, 0])
25
Node777: 1189 - 1206
        upstream: {N4(1186, 1188): 25}
        downstream: {N2(1207, 1207): 25}
        specimens: {137, 267, 397, 405, 279, 409, 154, 155, 159, 287, 417, 422, 423, 300, 307, 191, 320, 332, 346, 219, 226, 228, 236, 370, 371}
dict_keys([N4(1186, 1188)])
dict_values([25])
25
Node777: 1212 - 1215
        upstream: {N1(1208, 1211): 117, N3(1209, 1211): 44}
        downstream: {N1(1216, 1217): 144, N3(1216, 1225): 17}
        specimens: {1, 2, 4, 5, 10, 17, 22, 23, 24, 26, 28, 31, 33, 34, 37, 38, 43, 47, 50, 51, 54, 57, 62, 64, 65, 66, 70, 72, 73, 74, 76, 85, 88, 89, 92, 93, 99, 100, 101, 102, 103, 104, 105, 110, 112, 118, 120, 122, 126,

dict_keys([N1(1502, 1502), N4(1498, 1502), N-1(None, None)])
dict_values([291, 1, 0])
292
Node777: 1503 - 1503
        upstream: {N1(1502, 1502): 291, N4(1498, 1502): 1}
        downstream: {N1(1504, 1504): 277, N4(1504, 1504): 9, N5(1504, 1504): 5}
        specimens: {1, 2, 3, 4, 5, 6, 8, 9, 11, 12, 14, 16, 20, 21, 23, 24, 25, 26, 31, 32, 33, 34, 37, 38, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 52, 53, 54, 55, 56, 59, 60, 61, 63, 65, 68, 69, 70, 71, 72, 74, 75, 76, 78, 79, 80, 81, 82, 83, 84, 85, 86, 88, 89, 91, 93, 95, 96, 97, 99, 101, 103, 104, 106, 107, 108, 109, 110, 111, 113, 116, 117, 118, 119, 120, 121, 122, 123, 126, 127, 128, 130, 131, 132, 133, 134, 135, 136, 138, 139, 140, 141, 144, 145, 146, 147, 148, 149, 150, 151, 156, 157, 160, 161, 167, 168, 169, 170, 171, 172, 173, 176, 177, 178, 180, 181, 182, 183, 184, 185, 186, 187, 189, 190, 193, 194, 195, 196, 197, 200, 201, 204, 205, 207, 209, 210, 213, 214, 218, 220, 222, 225, 227, 228, 229, 234, 235, 238, 239, 240, 242, 244, 24

In [26]:
len(summary3) ## Order of nodes does matter here! HaploBlocker output: 1887

1865

### Everything below does not work currently. Some operations on empty sets lead to crashes. 
#### We should clean up upstream/downstream to not display 0 transition cases. del did not work directly

### Simple-merge / Cross-merge runs

In [27]:
window_cluster = blist([node for window in unique_signatures for node in window.values()])  # think about referencing and deletion
len(window_cluster)

for i in range(10):
    window_cluster = simple_merge(window_cluster)
    window_cluster = split_groups(window_cluster)
    print(len(window_cluster))


KeyError: set()

In [ ]:
len(window_cluster)

### Neglect nodes runs

In [ ]:
for i in range(10):
    window_cluster = neglect_nodes(window_cluster)
    window_cluster = simple_merge(window_cluster)
    window_cluster = split_groups(window_cluster)
    window_cluster = simple_merge(window_cluster)

    print(len(window_cluster))


In [ ]:
len(window_cluster)